### ISSUE TITLE
Use PostgreSQL's default permissions system to eliminate the need for the VZ & Moped permission ETLs

### ISSUE THREAD

**Issue by frankhereford** [cityofaustin/atd-data-tech#12532]
@johnclary wrote [in slack](https://austininnovation.slack.com/archives/CHZE6BC6L/p1686258523102909):

> this was what i had in mind for keeping read replica permissions always-up-to-date. theoretically you could set the default permissions for the read replica user group at the schema level https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html

This issue is to track implementing this system to eliminate the need for our two ETLs which maintain database permissions on the main RDS server so that the read replica remains usable. This will eliminate a prefect 1 ETL too, as a welcome side effect.

**Comment by johnclary** [comment]
@patrickm02L this needs to get prioritized this week.

**Comment by frankhereford** [comment]
The following SQL was executed first in the staging database on the the VZ DB server, and then the same SQL was run in production.

```
GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA public TO staff;
GRANT SELECT, USAGE ON ALL SEQUENCES IN SCHEMA public TO staff;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA public TO staff;

GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA hdb_catalog TO staff;
GRANT SELECT, USAGE ON ALL SEQUENCES IN SCHEMA hdb_catalog TO staff;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA hdb_catalog TO staff;

GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA hdb_views TO staff;
GRANT SELECT, USAGE ON ALL SEQUENCES IN SCHEMA hdb_views TO staff;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA hdb_views TO staff;



ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT, INSERT, UPDATE, DELETE ON tables TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT, USAGE ON SEQUENCES TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT EXECUTE ON FUNCTIONS TO staff;

ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT SELECT, INSERT, UPDATE, DELETE ON tables TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT SELECT, USAGE ON SEQUENCES TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT EXECUTE ON FUNCTIONS TO staff;

ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_views GRANT SELECT, INSERT, UPDATE, DELETE ON tables TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_views GRANT SELECT, USAGE ON SEQUENCES TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_views GRANT EXECUTE ON FUNCTIONS TO staff;
```

**Comment by johnclary** [comment]
nice! 🙌

**Comment by patrickm02L** [comment]
❤️  Thank You

**Comment by johnclary** [comment]
I just ran these same commands on Moped staging and prod—with the exception of the `hdb_view` schema which I think predates Hasura v2.x.

```sql
GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA public TO staff;
GRANT SELECT, USAGE ON ALL SEQUENCES IN SCHEMA public TO staff;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA public TO staff;

GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA hdb_catalog TO staff;
GRANT SELECT, USAGE ON ALL SEQUENCES IN SCHEMA hdb_catalog TO staff;
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA hdb_catalog TO staff;

ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT, INSERT, UPDATE, DELETE ON tables TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT, USAGE ON SEQUENCES TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT EXECUTE ON FUNCTIONS TO staff;

ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT SELECT, INSERT, UPDATE, DELETE ON tables TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT SELECT, USAGE ON SEQUENCES TO staff;
ALTER DEFAULT PRIVILEGES IN SCHEMA hdb_catalog GRANT EXECUTE ON FUNCTIONS TO staff;
```

thank you Frank!!!

**Comment by frankhereford** [comment]
@johnclary I'm going to reopen this issue; it looks like this may not be working as we hoped. I got this today:

```
pg_dump: error: query failed: ERROR:  permission denied for table council_districts
pg_dump: error: query was: LOCK TABLE public.council_districts IN ACCESS SHARE MODE
```

**Comment by johnclary** [comment]
Dang! Well, [this Moped PR](https://github.com/cityofaustin/atd-moped/pull/1083) will provide another testing an opportunity. For now I'm going to resist the urge to test this locally.

**Comment by johnclary** [comment]
@frankhereford i've retested this by querying the new `moped_proj_work_activity_status` on the moped read replica as well as a handful of test cases locally. i think we're all good on the Moped side. lmk how VZ is looking!

**Comment by frankhereford** [comment]
@chiaberry mentioned this issue to me today over slack, and it prompted me to give it another try with some testing. Thanks Chia!! 

I (re-)applied the default permission queries and did a test where I was able to successfully replicate the DB after adding a new table. 

I'm not sure what I did right this time that I didn't do last time .. maybe it stemmed from using the `postgres` role instead of the `vze` one. But it is working so I'm going to close this out until and unless more information surfaces. 🤞
